### PR TITLE
Add Mysql variables to collector

### PR DESCRIPTION
**Pull request by eochoa260**
This change adds output of `SHOW VARIABLES` query to the mysql collector to solve issue #555.  Tested locally with the text analyzer and it's working as expected.

### CODE REVIEW

File: pkg/collect/mysql.go [hunk 1]
```diff
@@ -27,6 +27,38 @@ func Mysql(c *Collector, databaseCollector *troubleshootv1beta2.Database) (Colle
 			databaseConnection.IsConnected = true
 			databaseConnection.Version = version
 		}
+
+		rows, err := db.Query("SHOW VARIABLES")
```
Comment (rocktavious) [inline pkg/collect/mysql.go]: This is more for replicated - but should this be "optional"?
If so how do we do that?

File: pkg/collect/mysql.go [hunk 2]
```diff
@@ -27,6 +27,38 @@ func Mysql(c *Collector, databaseCollector *troubleshootv1beta2.Database) (Colle
...
+		scanArgs := make([]interface{}, len(values))
+		for i := range values {
+			scanArgs[i] = &values[i]
+		}
+
+		variables := make(map[string]string)
+		for rows.Next() {
+			err = rows.Scan(scanArgs...)
+			if err != nil {
+				databaseConnection.Error = err.Error()
```
Comment (rocktavious) [inline pkg/collect/mysql.go:38]: we should exit the loop here with `break`

File: pkg/collect/mysql.go (hunk 1 above)
Comment (rocktavious) [reply pkg/collect/mysql.go]: Also @eochoa260 - i think `rows` is an object that we need to defer cleanup on
```
defer rows.Close()
```

File: pkg/collect/mysql.go [hunk 3]
```diff
@@ -27,6 +27,38 @@ func Mysql(c *Collector, databaseCollector *troubleshootv1beta2.Database) (Colle
...
+
+		rows, err := db.Query("SHOW VARIABLES")
+		if err != nil {
+			databaseConnection.Error = err.Error()
+		}
+		columns, err := rows.Columns()
+		if err != nil {
+			databaseConnection.Error = err.Error()
+		}
+		values := make([]sql.RawBytes, len(columns))
```
Comment (rocktavious) [inline pkg/collect/mysql.go]: I feel like we could clean this up.  I think `.Scan` can probably take a `map[string]interface{}`.

I'll do some tests tonight and report back.

**Review by rocktavious — CHANGES_REQUESTED** [review]
Just a few things to take care of.  Good work overall.

**Review by rocktavious — DISMISSED** [review]
LGTM now

File: pkg/collect/mysql.go [hunk 4]
```diff
@@ -20,6 +20,8 @@ func Mysql(c *Collector, databaseCollector *troubleshootv1beta2.Database) (Colle
 	} else {
 		query := `select version()`
 		row := db.QueryRow(query)
+		defer db.Close()
```
Comment (rocktavious) [inline pkg/collect/mysql.go]: @eochoa260 - move this before the `QueryRow` to ensure its hooked up incase something crashes the deferred call will still happen and clean up

File: pkg/collect/mysql.go [hunk 5]
```diff
@@ -15,18 +15,43 @@ func Mysql(c *Collector, databaseCollector *troubleshootv1beta2.Database) (Colle
...
 		}
+
+		rows, err := db.Query("SHOW VARIABLES")
+		defer rows.Close()
+
+		if err != nil {
+			databaseConnection.Error = err.Error()
+		}
+
+		if err != nil {
```
Comment (divolgin) [inline pkg/collect/mysql.go]: Line 36 is already doing this check

File: pkg/collect/mysql.go [hunk 6]
```diff
@@ -15,18 +15,43 @@ func Mysql(c *Collector, databaseCollector *troubleshootv1beta2.Database) (Colle
...
 		version := ""
 		if err := row.Scan(&version); err != nil {
 			databaseConnection.Error = err.Error()
 		} else {
 			databaseConnection.IsConnected = true
 			databaseConnection.Version = version
 		}
+
+		rows, err := db.Query("SHOW VARIABLES")
+		defer rows.Close()
```
Comment (divolgin) [inline pkg/collect/mysql.go]: `defer rows.Close()` should happen after error check

File: pkg/collect/mysql.go [hunk 7]
```diff
@@ -15,18 +15,43 @@ func Mysql(c *Collector, databaseCollector *troubleshootv1beta2.Database) (Colle
...
+		if err != nil {
+			databaseConnection.Error = err.Error()
+		}
+
+		if err != nil {
+			databaseConnection.Error = err.Error()
+		}
+
+		variables := map[string]string{}
+		for rows.Next() {
```
Comment (divolgin) [inline pkg/collect/mysql.go]: none of this should be executed if there is an error

File: pkg/collect/mysql.go (hunk 5 above)
Comment (eochoa260) [reply pkg/collect/mysql.go]: Oh yep, cleaned up some code here and must have left that one in there. Removing it now.

**Comment by divolgin** [comment]
I think this is a great addition, but variables may contain sensitive data (like `report_password` below?).

```
mysql> SHOW VARIABLES like "%pass%";
+----------------------------------------------+-----------------+
| Variable_name                                | Value           |
+----------------------------------------------+-----------------+
| caching_sha2_password_auto_generate_rsa_keys | ON              |
| caching_sha2_password_digest_rounds          | 5000            |
| caching_sha2_password_private_key_path       | private_key.pem |
| caching_sha2_password_public_key_path        | public_key.pem  |
| default_password_lifetime                    | 0               |
| disconnect_on_expired_password               | ON              |
| generated_random_password_length             | 20              |
| mysql_native_password_proxy_users            | OFF             |
| password_history                             | 0               |
| password_require_current                     | OFF             |
| password_reuse_interval                      | 0               |
| report_password                              |                 |
| sha256_password_auto_generate_rsa_keys       | ON              |
| sha256_password_private_key_path             | private_key.pem |
| sha256_password_proxy_users                  | OFF             |
| sha256_password_public_key_path              | public_key.pem  |
+----------------------------------------------+-----------------+
16 rows in set (0.00 sec)
```

The default redactors may not pick this up.  One option is to make this data optional, for example

```
  collectors:
    - mysql:
        collectorName: mysql
        uri: ...
        includeVariables: true
```

**Comment by rocktavious** [comment]
@divolgin 
Thats a great point.  Could we also configure it so you have to specify which variable names you want to pull back?  We aren't 100% up to speed on how to add extra configuration like this from the collector definition the user fills out.

Maybe something like
```
  collectors:
    - mysql:
        collectorName: mysql
        uri: ...
        withVariables:
        - character_set_server
        - collation_server
        - innodb_file_format
```

File: pkg/collect/database_shared.go [hunk 8]
```diff
@@ -1,7 +1,13 @@
 package collect
 
+type MySQLVariable struct {
```
Comment (divolgin) [inline pkg/collect/database_shared.go]: There is also no need for this type

File: pkg/collect/mysql.go [hunk 9]
```diff
@@ -18,15 +18,37 @@ func Mysql(c *Collector, databaseCollector *troubleshootv1beta2.Database) (Colle
...
+		rows, err := db.Query("SHOW VARIABLES")
+
+		if err != nil {
+			databaseConnection.Error = err.Error()
+		} else {
+			defer rows.Close()
+
+			variables := map[string]string{}
+			for rows.Next() {
+				variable := MySQLVariable{}
```
Comment (divolgin) [inline pkg/collect/mysql.go]: This should just be `var key, value string`

Comment (eochoa260) [reply pkg/collect/mysql.go]: Oh nice, thanks for the suggestion.  Just made that update.

**Comment by eochoa260** [comment]
> I think this is a great addition, but variables may contain sensitive data (like `report_password` below?).
> 
> ```
> mysql> SHOW VARIABLES like "%pass%";

> +----------------------------------------------+-----------------+
> | Variable_name                                | Value           |
> +----------------------------------------------+-----------------+
> | caching_sha2_password_auto_generate_rsa_keys | ON              |
> | caching_sha2_password_digest_rounds          | 5000            |
> | caching_sha2_password_private_key_path       | private_key.pem |
> | caching_sha2_password_public_key_path        | public_key.pem  |
> | default_password_lifetime                    | 0               |
> | disconnect_on_expired_password               | ON              |
> | generated_random_password_length             | 20              |
> | mysql_native_password_proxy_users            | OFF             |
> | password_history                             | 0               |
> | password_require_current                     | OFF             |
> | password_reuse_interval                      | 0               |
> | report_password                              |                 |
> | sha256_password_auto_generate_rsa_keys       | ON              |
> | sha256_password_private_key_path             | private_key.pem |
> | sha256_password_proxy_users                  | OFF             |
> | sha256_password_public_key_path              | public_key.pem  |
> +----------------------------------------------+-----------------+
> 16 rows in set (0.00 sec)
> ```
> 
> The default redactors may not pick this up. One option is to make this data optional, for example
> 
> ```
>   collectors:
>     - mysql:
>         collectorName: mysql
>         uri: ...
>         includeVariables: true
> ```

@divolgin I made an update to only pull mysql variables that are in defined as parameters in the preflight manifest, e.g. - 
```spec:
  collectors:
    - mysql:
        collectorName: mysql
        uri: 'root:my-secret-pw@tcp(localhost:3306)/mysql'
        parameters:
          - character_set_server
          - collation_server
          - init_connect
          - innodb_file_format
          - innodb_large_prefix
          - innodb_strict_mode
          - log_bin_trust_function_creators
```
Look forward to any feedback you have on this though.

**Review by divolgin — APPROVED** [review]

**Comment by emosbaugh** [comment]
I tested this with the provided yaml. It works with the following output. Merging.

```
$ cat support-bundle-2022-05-04T16_51_16/mysql/mysql.json 
{"isConnected":true,"version":"8.0.29","variables":{"character_set_server":"utf8mb4","collation_server":"utf8mb4_0900_ai_ci","init_connect":"","innodb_strict_mode":"ON","log_bin_trust_function_creators":"OFF"}}
```

**Pull request merged**